### PR TITLE
DAOS-8465 test: Loop to fill and destroy container and pool.

### PR DESCRIPTION
**Pull request by dinghwah**
Description:  repeat following steps:
   (1)Create Pool and Container.
   (2)Fill the pool with random block data size, verify return code is as expected
      DER_NOSPACE -1007 when no more data can be written to the container.
   (3)destroy the container

Skip-unit-tests: true
Test-tag: fill_cont_pool_stress
Signed-off-by: Ding Ho ding-hwa.ho@intel.com

### CODE REVIEW

File: src/tests/ftest/container/fill_destroy_loop.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        space_avail_to_write = True
+        while space_avail_to_write:
+            dkey = get_random_bytes(dkey_size)
+            akey = get_random_bytes(akey_size)
+            try:
+                written_byte += data_size
+                self.d_log.debug("--{0} writing cont-obj {1}  byte data, total {2}..".format(
+                    write_count, data_size, written_byte))
+                container.written_data.append(TestContainerData(False))
+                container.written_data[-1].write_record(
```
Comment (shimizukko) [inline src/tests/ftest/container/fill_destroy_loop.py:49]: I think it's better to use the higher-level API, `TestContainer.write_objects_wo_failon()`, to write data. It'll eventually call the `write_record()` and you don't need to prepare the `dkey` and `akey`.

**Review by shimizukko — DISMISSED** [review]

File: src/tests/ftest/container/fill_destroy_loop.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
+#!/usr/bin/python3
```
Comment (daltonbohning) [inline src/tests/ftest/container/fill_destroy_loop.py:1]: Minor - shebang should not be added

File: src/tests/ftest/container/fill_destroy_loop.py (hunk 1 above)
Comment (daltonbohning) [reply src/tests/ftest/container/fill_destroy_loop.py:49]: I'm not familiar with `TestContainer.write_objects_wo_failon()`, but +1 if it simplifies and reduces redundant code

File: src/tests/ftest/container/fill_destroy_loop.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+            repeat following steps:
+            (1)Create Pool and Container.
+            (2)Fill the pool with random block data size, verify return code is as expected
+               DER_NOSPACE -1007 when no more data can be written to the container.
+            (3)destroy the container
+
+        :avocado: tags=all,full_regression
+        :avocado: tags=hw,medium
+        :avocado: tags=container,pool
+        :avocado: tags=fill_cont_pool_stress
```
Comment (daltonbohning) [inline src/tests/ftest/container/fill_destroy_loop.py:82]: Minor - can we please add a unique tag corresponding to the method name?
```suggestion
        :avocado: tags=fill_cont_pool_stress,test_fill_destroy_cont_loop
```

File: src/tests/ftest/container/fill_destroy_loop.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        """
+
+        testloop = self.params.get("testloop", "/run/pool/*")
+
+        # create pool
+        self.add_pool()
+
+        for test_loop in range(1, testloop+1):
```
Comment (daltonbohning) [inline src/tests/ftest/container/fill_destroy_loop.py:90]: Minor - a silent/future flake warning wants spaces
```suggestion
        for test_loop in range(1, testloop + 1):
```

File: src/tests/ftest/container/fill_destroy_loop.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+            # query the pool and get free space before write
+            self.log.info("==>Starting test loop: %s ...", test_loop)
+            self.pool.set_query_data()
+            self.log.info("--(1)Pool Query before write:\n"
+                "--Pool %s query data: %s\n", self.pool.uuid, self.pool.query_data)
```
Comment (daltonbohning) [inline src/tests/ftest/container/fill_destroy_loop.py:95]: Minor - style - better to align (future pylint/flake will check for this)
```suggestion
            self.log.info(
                "--(1)Pool Query before write:\n"
                "--Pool %s query data: %s\n", self.pool.uuid, self.pool.query_data)
```
or
```suggestion
            self.log.info("--(1)Pool Query before write:\n"
                          "--Pool %s query data: %s\n", self.pool.uuid, self.pool.query_data)
```

File: src/tests/ftest/container/fill_destroy_loop.yaml [hunk 6]
```diff
@@ -0,0 +1,17 @@
+hosts:
+  test_servers: 3
+server_config:
+  name: daos_server
+timeout: 500
+pool:
+  name: daos_server
+  control_method: dmg
+  scm_size: 1G
```
Comment (daltonbohning) [inline src/tests/ftest/container/fill_destroy_loop.yaml:9]: Could we use `size` instead, and let DAOS figure out if it's SCM and/or NVMe based on the config? Or do we specifically only ever want SCM?

File: src/tests/ftest/container/fill_destroy_loop.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+                self.d_log.debug("--{0}wrote cont-obj, sz {1}".format(write_count, data_size))
+                write_count += 1
+            except DaosTestError as excep:
+                space_avail_to_write = False
+                if self.DER_NOSPACE in repr(excep):
+                    self.log.info("--(3)written_byte= %s, Der_no_space %s detected, pool is "
+                                  "unable for an additional %s byte object", written_byte,
+                                  self.DER_NOSPACE, data_size)
+                    free_space = self.pool.get_pool_free_space()
+                    self.log.info("--(4)free_space when pool is fulfill= %s", free_space)
```
Comment (daltonbohning) [inline src/tests/ftest/container/fill_destroy_loop.py:60]:
```suggestion
                    self.log.info("--(4)free_space when pool is full = %s", free_space)
```

File: src/tests/ftest/container/fill_destroy_loop.py [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+                else:
+                    self.fail("#Exception while writing object: {}".format(repr(excep)))
```
Comment (phender) [inline src/tests/ftest/container/fill_destroy_loop.py:62]: Let's use the methods available in the TestContainer class and akey_size, dkey_size, and data_size test yaml parameters:
```suggestion
        container = self.get_container(self.pool)

        # Select a random data size based on block_size
        new_data_size = random.randint(container.data_size.value * 0.5, container.data_size.value * 1.5)   # nosec
        container.data_size.update(new_data_size, "data_size")

        while(1):
            try:
                container.write_objects()
            except DaosTestError as error:
                if self.DER_NOSPACE in error:
                    self.log.info(
                        "--(3) Der_no_space %s detected, pool is unable for an additional %s byte object",
                        self.DER_NOSPACE, container.data_size.value)
                    free_space = self.pool.get_pool_free_space()
                    self.log.info("--(4)free_space when pool is fulfill= %s", free_space)
                    break
                else:
                    self.fail("#Exception while writing object: {}".format(repr(excep)))
```

File: src/tests/ftest/container/fill_destroy_loop.yaml [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+  testloop: 100
+container:
+  control_method: daos
+  properties: "rf:1"
+  obj_class: OC_RP_2G1
+  block_size: 1024000
```
Comment (phender) [inline src/tests/ftest/container/fill_destroy_loop.yaml:15]: This should be data_size:
```suggestion
  data_size: 1024000
```

**Review by phender — CHANGES_REQUESTED** [review]

File: src/tests/ftest/container/fill_destroy_loop.py (hunk 1 above)
Comment (dinghwah) [reply src/tests/ftest/container/fill_destroy_loop.py:49]: Thank you for the feedback.
However, this is a write container object until pool full, if we use container.write_objects_wo_failon(), I am seeing double exception when pool is full:
   "Object update returned non-zero. RC: -1007 During handling of the above exception, another exception occurred:"

So better to use low level api container.written_data.write_record() until pool is full and trigger the exception.

Comment (phender) [reply src/tests/ftest/container/fill_destroy_loop.py:37]: > I think it's better to use the higher-level API, `TestContainer.write_objects_wo_failon()`, to write data. It'll eventually call the `write_record()` and you don't need to prepare the `dkey` and `akey`.

@shimizukko I didn't know we added this method.  It's not really needed as all the `@failon()` does is change the exception raised to an `avocado.core.exceptions.TestFail` exception.  This can be caught instead of the `DaosTestError` exception.

File: src/tests/ftest/container/fill_destroy_loop.py [hunk 10]
```diff
@@ -25,42 +24,43 @@ class BoundaryPoolContainerSpace(TestWithServers):
...
-                    self.log.info("--(4)free_space when pool is fulfill= %s", free_space)
-                else:
-                    self.fail("#Exception while writing object: {}".format(repr(excep)))
+                container.write_objects()
+            except TestFail as excep:
+                # Uncomment following for debugging
+                # self.log.info("%s", repr(excep))
+                if self.DER_NOSPACE in str(excep):
+                    self.log.info(
+                        "--(3) Der_no_space %s detected, pool is unable for an additional"
```
Comment (daltonbohning) [inline src/tests/ftest/container/fill_destroy_loop.py:47]:
```suggestion
                        "--(3) DER_NOSPACE %s detected, pool is unable for an additional"
```

File: src/tests/ftest/container/fill_destroy_loop.py [hunk 11]
```diff
@@ -25,42 +24,43 @@ class BoundaryPoolContainerSpace(TestWithServers):
...
+        free_space_before = self.pool.get_pool_free_space()
+        self.log.info("--(4)%s bytes written when pool is full.", data_written)
+
+        # destroy container and check for free space increase
         container.destroy()
+        free_space_after = self.pool.get_pool_free_space()
+        self.log.info("--(5)pooll full, free_space before container delete= %s", free_space_before)
+        self.log.info("--(6)pooll full, free_space after  container deleted= %s", free_space_after)
+        if free_space_after <= free_space_before:
+            self.fail("Deleting container did not free up pool space.")
```
Comment (daltonbohning) [inline src/tests/ftest/container/fill_destroy_loop.py:63]: Do we need to wait for aggregation? I'm not sure if free space is immediately returned

**Review by daltonbohning — COMMENTED** [review]
LGTM, except for my question about aggregation

**Review by shimizukko — APPROVED** [review]

File: src/tests/ftest/container/fill_destroy_loop.py [hunk 12]
```diff
@@ -85,8 +85,9 @@ def test_fill_destroy_cont_loop(self):
         :avocado: tags=fill_cont_pool_stress,test_fill_destroy_cont_loop
         """
         testloop = self.params.get("testloop", "/run/pool/*")
-        # create pool
+        # create pool and enable the aggregation
         self.add_pool()
+        self.pool.set_property("reclaim", "time")
```
Comment (daltonbohning) [inline src/tests/ftest/container/fill_destroy_loop.py:90]: I think aggregation is enabled by default (maybe I'm wrong). But also, my previous comment was about whether or not we need to wait for aggregation to start before checking `if free_space_after <= free_space_before:` to avoid intermittent failures. (again, I might be wrong about this)

**Comment by kprantis** [comment]
PS, please install the team's githooks as Jenkins is complaining about for the future. :)

File: src/tests/ftest/container/fill_destroy_loop.yaml [hunk 13]
```diff
@@ -0,0 +1,25 @@
...
+    bdev_class: nvme
+    bdev_list: ["0000:81:00.0"]
+    scm_class: dcpm
+    scm_list: ["/dev/pmem0"]
+timeout: 1200
+pool:
+  name: daos_server
+  control_method: dmg
+  scm_size: 1G
+  nvme_size: 26G
```
[the remark this answers is not in the document]
Comment (dinghwah) [reply src/tests/ftest/container/fill_destroy_loop.yaml:15]: From this pr (pr-9859/16): I only specified size (without scm_size and nvme_size) and seeing following error:
/usr/bin/dmg -o /etc/daos/daos_control.yml -d -j pool create --group=jenkins --label=TestLabel_1 --size=1G --user=jenkins' finished with 1 after 0.16453170776367188s
2022-08-16 21:37:39,675 dmg_utils        L0542 ERROR| pool create failed: server: code = 604 description = "requested SCM capacity is too small (min 16 MiB per target)"

[the remark this answers is not in the document]
Comment (daltonbohning) [reply src/tests/ftest/container/fill_destroy_loop.yaml:15]: Right. That's because the _**total SCM + NVMe size**_ is 1G. Whereas if you specify `scm_size=1G`, the total SCM size is `1G * num_engines`. If you use `size`, you probably need something like 50G+

**Comment by dinghwah** [comment]
Noted:
I reserved shared cluster-A with NVME for more testing, and found out that 1 server the minimum pool size is ~10GB, test to fill pool and repeat 100 times will consume 3+ hours.  So I use 1 server and scm only. 
We can add more testcase with !mux for more servers and NVME if it is require.

File: src/tests/ftest/container/fill_destroy_loop.py [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+
+        :avocado: tags=all,full_regression
+        :avocado: tags=hw,small
+        :avocado: tags=container,pool
+        :avocado: tags=fill_cont_pool_stress,test_fill_destroy_cont_loop
+        """
+        testloop = self.params.get("testloop", "/run/pool/*")
+        # create pool and enable the aggregation
+        self.add_pool()
+        self.pool.set_property("reclaim", "time")
```
Comment (daltonbohning) [inline src/tests/ftest/container/fill_destroy_loop.py:100]: Since this is set upfront, I think we could also do this in the yaml:
```
pool:
  properties: reclaim:time
```

**Review by daltonbohning — APPROVED** [review]
LGTM

**Review by phender — APPROVED** [review]

**Review by shimizukko — APPROVED** [review]
